File: node/gravitate.mjs
const GRAVITY = 6.6740831e-6; //6.6740831e-11;

export function gravitate(entity, satellite) {
  if (entity.id === satellite.id) {
    return;
  }

 // if (entity.mass <= 0 || satellite.mass <= 0) {
 //   return;
 // }

  const distance = entity.distance(satellite);

  if (distance <= 0) {
    return;
  }

  const delta = entity.positionDeltaFrom(satellite);

  const gravitation = GRAVITY * ((entity.mass * satellite.mass) / (distance ** 2));

  if (gravitation === 0) {
    return;
  }

  const force = {
    x: (delta.x * gravitation) / distance,
    y: (delta.y * gravitation) / distance,
    z: (delta.z * gravitation) / distance
  };

  entity.add(force.x, force.y, force.z);
}

File: cpp/gravitate.cpp
#include "vector.h"
#include "entity.h"
#include "gravitate.h"

Vector3D gravitate (const Entity& entity, const Entity& satellite) {
  Vector3D force (0, 0, 0);
  
  if (entity.id == satellite.id) {
    return force;
  }
  
  long double distance = entity.position->distance(*satellite.position);
  
  if (distance <= 0) {
    return force;
  }
  
  Vector3D deltaFrom = *satellite.position - *entity.position;
  
  long double gravitation = GRAVITY * ((entity.mass * satellite.mass) / (distance * distance));
  
  if (gravitation == 0) {
    return force;
  }
  
  force.x = (deltaFrom.x * gravitation) / distance;
  force.y = (deltaFrom.y * gravitation) / distance;
  force.z = (deltaFrom.z * gravitation) / distance;
  
  return force;
}
